Sources/HummingbirdWebSocket/WebSocketHandler.swift

```diff
@@ -16,96 +16,120 @@ import Hummingbird
+                                        }
+                                    default:
+                                        break
+                                    }
+                                    if let frameSeq = frameSequence, frame.fin {
+                                        await webSocket.inbound.send(frameSeq.data)
+                                        frameSequence = nil
+                                    }
+                                } catch {
+                                    // catch errors while processing websocket frames so responding close message
+                                    // can be dealt with
+                                    let errorCode = WebSocketErrorCode(error)
+                                    try await self.close(code: errorCode, outbound: webSocket.outbound, context: context)
                                 }
-                            case .continuation:
-                                if var frameSeq = frameSequence {
-                                    frameSeq.append(frame)
-                                    frameSequence = frameSeq
+                            }
+                        }
+                        group.addTask {
+                            do {
+                                // handle websocket data and text
+                                try await handler(webSocket, context)
+                                try await self.close(code: .normalClosure, outbound: webSocket.outbound, context: context)
+                            } catch {
+                                if self.type == .server {
+                                    let errorCode = WebSocketErrorCode.unexpectedServerError
+                                    try await self.close(code: errorCode, outbound: webSocket.outbound, context: context)
                                 } else {
-                                    try await self.close(code: .protocolError, outbound: outbound, context: context)
+                                    try await asyncChannel.channel.close(mode: .input)
                                 }
-                            default:
-                                break
-                            }
-                            if let frameSeq = frameSequence, frame.fin {
-                                await webSocketHandlerInbound.send(frameSeq.data)
-                                frameSequence = nil
                             }
                         }
+                        try await group.next()
+                        webSocket.inbound.finish()
                     }
-                    group.addTask {
-                        // handle websocket data and text
-                        try await handler(webSocketHandlerInbound, webSocketHandlerOutbound, context)
-                        try await self.close(code: .normalClosure, outbound: outbound, context: context)
+                } onGracefulShutdown: {
+                    Task {
```

Comment:
Is there a way around an unstructured task here?

Reply:
Unfortunately not. I can't close either the inbound stream or outbound writer as they are needed for the close handshake. It's a recurring issue with graceful shutdown handlers, where something more complex is need to initiate shutdown.